Better defaults for model inference

kaixi-wang opened the pull request:
## What changes are proposed in this pull request?

Improve inference performance for torch models by improving defaults. 
- Add/expose a new env var for model inference related batch_size `FIFTYONE_DEFAULT_MODEL_BATCH_SIZE`
- to differentiate from default_batch_size used by the sdk for database operations
- Expose `prefetch_factor` via env var `FIFTYONE_DEFAULT_DATALOADER_PREFETCH_FACTOR` and as an arg to pass to torch dataloaders
- cap number of recommended default workers to 4

## How is this patch tested? If it is not, please explain why.

Ran benchmarks for several models with apply_model, setting batch_size and num_workers on GPU and high CPU compute:

### With A100 GPU:
<img width="1462" height="1072" alt="image" src="https://github.com/user-attachments/assets/e06837a0-148d-4568-89bc-523c93e0068a" />
<img width="1399" height="1073" alt="image" src="https://github.com/user-attachments/assets/32847026-b4ba-4937-9cde-a52e8cc0f577" />
<img width="1430" height="1076" alt="image" src="https://github.com/user-attachments/assets/323cc37f-d1d1-4b74-8160-6bc2f247606e" />

### With compute-optimized 48 cpu:

<img width="1393" height="1063" alt="image" src="https://github.com/user-attachments/assets/a931e10b-6fb4-438e-9801-1f20293d1ef4" />
<img width="1413" height="1080" alt="image" src="https://github.com/user-attachments/assets/50e19b26-6b93-4256-819b-978d4914cdcf" />


## Release Notes

### Is this a user-facing change that should be mentioned in the release notes?

<!--
Please fill in relevant options below with an "x", or by clicking the checkboxes
after submitting this pull request. Example:
-   [x] Selected option
-->

-   [ ] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Two new config options: default model batch size and default dataloader prefetch factor.
  * Prefetch control parameter added to model inference and embedding workflows.

* **Improvements**
  * More robust CPU detection across environments, used for worker defaults.
  * Smarter default worker recommendation with a capped maximum for stability.

* **Tests**
  * Unit tests updated to use the new CPU-count helper and expanded CPU-count coverage.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

brimoor commented on fiftyone/core/utils.py:
```diff
@@ -2870,6 +2870,39 @@ def get_multiprocessing_context():
     return multiprocessing.get_context()
 
 
+def get_cpu_count():
```
@swheaton is proposing to add a `get_cpu_count()` utility in https://github.com/voxel51/fiftyone/pull/6290, so these two implementations should be reconciled and we should ship a single thing that does whatever combo of things we consider best practice.

kaixi-wang replied:
Oh thanks for calling that out. Didn't know Stuart had another implementation.

brimoor commented on fiftyone/core/models.py:
```diff
@@ -1994,18 +2003,20 @@ def _patch_collate_fn(batch):
 
 
-def _parse_batch_size(batch_size, model, use_data_loader):
+def _parse_batch_size_and_num_workers(*, batch_size, model, num_workers):
```
The reason the existing implementation handles `batch_size` default separately from `num_workers` is that `apply_model()` supports non-Torch models and `batch_size` is meaningful for all models, whereas `num_workers` is only relevant for Torch models that use data loaders.

With this change, running inference with a Tensorflow model will raise a "this method requires 'torch' to be installed" error:

```py
import fiftyone as fo
import fiftyone.zoo as foz

dataset = foz.load_zoo_dataset("quickstart")
model = foz.load_zoo_model("efficientdet-d6-1280-coco-tf2")

# Raises "torch" not installed error
dataset.apply_model(model, label_field="preds")
```

because `_parse_batch_size_and_num_workers()` is executed and now imports `fiftyone.utils.torch`.

Unless/until `torch` is a mandatory requirement of `fiftyone`, I don't think that should happen.

jacobsela reviewed (changes requested):
This PR is doing a lot at once, some good some I don't fully understand. If we limit the PR to the good we can merge quickly and handle the rest in other PRs IMO.

Good:
- introducing `prefects_factor` arg. This is great. Note that the prefetch factor increases memory usage linearly. (It should still be technically better than more workers because we avoid the overhead of new processes).

Can go either way:
- cap default worker usage to 4 - generally fine, if people want more they can chuck in whatever value they want. I don't agree with the reasoning given though, there are envs / models / circumstances that benefit from higher num_workers.

Needs explanation:
- global default batch size > 1 - I agree with Brian's comments.
- changing default thread count - presumably threaded operations are I/IO heavy. We don't necessarily want to cap the number of threads to be the number of CPU cores because we may want to have > num cpu cores pending operations at once.

kaixi-wang commented:
> global default batch size > 1 - I agree with Brian's comments.

In my latest, I updated it to 1 as the default. The goal is just to have a separate env var for it because if someone sets the DEFAULT_BATCH_SIZE to 1, then that affects mongo operations and we should definitely not do that

> changing default thread count - presumably threaded operations are I/IO heavy. We don't necessarily want to cap the number of threads to be the number of CPU cores because we may want to have > num cpu cores pending operations at once.

Removed it. I was just going by the docs since they sound reasonable (https://docs.pytorch.org/docs/stable/notes/multiprocessing.html#avoid-cpu-oversubscription). However,  I didn't really observe any difference with setting it, and I believe the OOM issues we've seen with docker deployments is the inaccurate cpu_count() that is addressed with get_cpu_counts()
